File: Grader/a65_q2b_arch.py
import sys

def input():
    return sys.stdin.readline()

def perform(l, r):
    if l >= r:
        return 0
    if dp[l][r] == sys.maxsize:
        res = 0
        #use pair l, r
        res = perform(l+1, r-1) + v[l] * v[r]
        #not use pair l, r => calculate all possible from l to r
        for i in range(l+1, r):
            res = max(res, perform(l, i) + perform(i+1, r))
        dp[l][r] = res
    return dp[l][r]

n = int(input())
dp = [[sys.maxsize] * 501 for _ in range(501)]
v = [0] + list(map(int, input().split()))

ans = perform(1, n)
print(ans)

File: Grader/a65_q2b_arch.cpp
#include <bits/stdc++.h>
using namespace std;

int n, num;
vector<vector<int>> memo(501, vector<int>(501, INT_MAX));
vector<int> v;

int perform(int l, int r){
    if(l >= r) return 0;
    int res = memo[l][r];

    if(res == INT_MAX){
        res = 0;
        //res = max( res, perform(l+1,r), perform(l,r-1) )
        res = max(res, perform(l+1, r));
        res = max(res, perform(l, r-1));
        for(int i=l+1; i<r+1; ++i){
            auto val = v[l] * v[i] + ( perform(l+1, i-1) + perform(i+1, r) );
            res = max(res, val);
        }   
    }
    memo[l][r] = res;
    return res;
}
int main() {
//  testcases 
//  8
//  3 9 -1 4 -5 5 2 7
//    cin >> n;
    v.resize(8);
    // for(int i=0; i<n; ++i){
    //     cin >> v[i];
    // }
    v = {3, 9, -1, 4, -5, 5, 2, 7};
    int ans = perform(0, 8-1);
    cout << ans;
}